### PR TITLE
Add non-strict backward patching helper

### PR DESCRIPTION
**Pull request by tugsbayasgalan**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #181012

Add torch.compiler._patch_engine_backward() so non-strict make_fx tracing installs the same stack trace, seq_nr, and autograd_backward preservation hooks for loss.backward() that we already rely on for torch.autograd.grad(). This extends the tagging path to backward()-driven traces without changing the eager autograd behavior outside the helper.

Keep torch.compiler._patch_autograd_grad() for existing users, but deprecate it with a once-only warning and leave a TODO to remove it once TorchTitan stops depending on the helper. The non-strict activation checkpointing tests now cover both the guard and the metadata leak boundary for backward().

Authored with Codex, continuing earlier Claude-assisted work.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @Lucaskabela @jataylo @azahed98

### CODE REVIEW

File: torch/compiler/__init__.py [hunk 1]
```diff
@@ -524,10 +524,17 @@ def _patch_autograd_grad():
     import functools
 
     import torch.autograd
+    from torch._dynamo.utils import warn_once
     from torch._functorch._aot_autograd.logging_utils import (
         setup_stacktrace_preservation_hooks_from_tensors,
     )
 
+    warn_once(
+        "torch.compiler._patch_autograd_grad() is deprecated; "
```
Comment (SherlockNoMad) [inline torch/compiler/__init__.py:533]: is this just a rename of _patch_autograd_grad?

**Review by SherlockNoMad — APPROVED** [review]

**Comment by SherlockNoMad** [comment]
I think we should make minimal_make_fx tracer an annex project when API stabilizes.
